src/documents/templates/ExtractItem.ts

```diff
@@ -0,0 +1,113 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from "vscode";
+import { IActionContext, IAzureUserInput } from "vscode-azureextensionui";
+import { ObjectValue } from "../../language/json/JSON";
+import { DeploymentTemplateDoc } from "./DeploymentTemplateDoc";
+import { InsertItem } from "./insertItem";
+export class ExtractItem {
+    // tslint:disable-next-line:no-empty
+    constructor(private ui: IAzureUserInput) {
+    }
+
+    public async extractParameter(editor: vscode.TextEditor, template: DeploymentTemplateDoc, context: IActionContext): Promise<void> {
+        let selection = this.expandSelection(editor.selection, editor.document, template, editor);
+        let selectedText = editor.document.getText(selection);
+        let name = await this.ui.showInputBox({ prompt: "Name of parameter?" });
+        const leaveEmpty = "Press 'Enter' if you do not want to add a description.";
+        let description = await this.ui.showInputBox({ prompt: "Description?", placeHolder: leaveEmpty });
+        const insertText = `[parameters('${name}')]`;
+        const texts = this.fixExtractTexts(selectedText, insertText, selection, template, editor);
+        let topLevel = this.getTopLevel(template, selection);
+        await editor.edit(builder => builder.replace(selection, texts[1]), { undoStopBefore: true, undoStopAfter: false });
+        await new InsertItem(this.ui).insertParameterWithDefaultValue(topLevel, editor, context, name, texts[0], description, { undoStopBefore: false, undoStopAfter: true });
```

Comment:
nit: After extracting, we should ask the editor to format the inserted range.

Reply:
Example output before formatting:
```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.2.2.3",
    "resources": [
        {
            "type": "Microsoft.Compute/virtualMachines/extensions",
            "apiVersion": "[parameters('abc')]",
            "name": "[concat(parameters('vmName'),copyIndex(1),'/dscext')]",
            "dependsOn": [
                // >>> SET CURSOR ON FOLLOWING LINE BETWEEN "r" and "esourceId"
                "[resourceId('Microsoft.Compute/virtualMachines/extensions', concat(parameters('vmName'),copyIndex(1),'/dscext'))]"
            ]
      ,
    "parameters": {
        "abc": {
            "type": "string",
            "defaultValue": "2019-12-01"
        }
    }  }
    ]
}
```

Reply:
This problem was fixed when fixing another bug. There are still problems with indentation for inner scoped nested deployments though. Is that still a bug that needs to be fixed?
Can you explain more on how to ask the editor to format the inserted range?

Reply:
I think I fixed the indenting issues, so we don't need to worry about this for now.